File: neural_network/components/network.py
from typing import List
import math
import random

import matplotlib.pyplot as plt
import numpy as np
import networkx as nx

from neural_network.functions import TransferFunction
from neural_network.functions import ReLU
from neural_network.functions import Softmax
from neural_network.functions import MSELoss

from .neuron import Neuron
from .edge import Edge
from .layer import Layer


class Network:
    """Class to represent the whole `Network`
    """

    def __init__(self, num_features: int, num_hidden_layers: int,
                 neuron_counts: List[int], regression: bool = False,
                 leak: float = 0.01, learning_rate: float = 0.01,
                 num_classes: int = 2, adaptive: bool = False,
                 gamma: float = 0.9, he_weights: bool = False):
        """Constructor method

        Parameters
        ----------
        num_features : int
            The number of features for the network
        num_hidden_layers : int
            The total number of hidden `Layers` in the `Network`
        neuron_counts : List[int]
            A list of numbers of `Neurons` for each hidden `Layer`
        regression : bool
            Whether we are performing regression or not (if `False` we are
            performing classification)
        leak : float
            The leak rate of LeakyReLU
        learning_rate : float
            The learning rate of the network
        num_classes : int
            The number of classes for the classification task
        adaptive : bool
            Whether we wish to have an adaptive learning rate or not
        gamma : float
            The adaptive learning rate parameter
        he_weights : bool
            Whether we wish to initialise the weights according to He or not
        """
        if not num_hidden_layers == len(neuron_counts):
            raise ValueError(f"neuron_counts ({len(neuron_counts)}) must have "
                             f"a length equal to num_hidden_layers "
                             f"({num_hidden_layers})")
        self._num_features = num_features
        self._num_hidden_layers = num_hidden_layers
        self._neuron_counts = neuron_counts
        self._regression = regression
        # If we are doing regression set num_classes to 1, no matter what the
        # input was
        if regression:
            num_classes = 1

        # Layers
        self._input_layer = Layer(0, num_features)
        self._hidden_layers = [Layer(i, neuron_counts[i - 1])
                               for i in range(1, num_hidden_layers + 1)]
        self._output_layer = Layer(num_hidden_layers + 1, num_classes)
        self._layers = ([self._input_layer] + self._hidden_layers
                        + [self._output_layer])

        # Edges
        self._edges = []
        for i, left_layer in enumerate(self._layers[:-1]):
            right_layer = self._layers[i + 1]
            layer_list = []
            for right_neuron in right_layer.get_neurons():
                edge_list = [Edge(left_neuron, right_neuron)
                             for left_neuron in left_layer.get_neurons()]
                layer_list.append(edge_list)
                if he_weights:
                    n = len(left_layer)
                    for edge in edge_list:
                        edge.set_weight(random.gauss(0.0,
                                                     math.sqrt(2 / n)))
            self._edges.append(layer_list)

        # Functions and other parameters
        self._transfer = TransferFunction()
        self._relu = ReLU(leak)
        if regression:
            self._mse_loss = MSELoss()
        else:
            self._softmax = Softmax()
        self._learning_rate = learning_rate
        self._adaptive = adaptive
        self._gamma = gamma

    def forward_pass_one_datapoint(self, x: np.array) -> List[float]:
        """Performs a forward pass for one datapoint, excluding the ground
        truth value. This method returns the predicted value in the final
        neuron in the output layer.

        Parameters
        ----------
        x : np.array
            The input value, with all features

        Returns
        -------
        List[float]
            The softmax probabilities of each class (for classification) or
            the predicted regression value (for regression)
        """
        if not len(x) == len(self._input_layer):
            raise ValueError(f"Number of features must match the number of "
                             f"neurons in the input layer ({len(x)} != "
                             f"{len(self._input_layer)})")

        # Input layer
        input_neurons = self._input_layer.get_neurons()
        for j, neuron in enumerate(input_neurons):
            neuron.set_value(x[j])

        # This is for the output layer
        z_output_layer = []
        # Hidden layers and output layer
        for left_layer in self._layers[:-1]:
            right_layer = self._layers[left_layer.get_id() + 1]

            if right_layer.get_id() != self._output_layer.get_id():
                for right_neuron in right_layer.get_neurons():
                    # Calculates the desired values for each neuron
                    z = self._calculate_pre_activated_value(left_layer,
                                                            right_neuron)
                    # Uses ReLU activation for the neuron
                    right_neuron.set_value(self._relu(z))
            else:
                if self._regression:
                    # We only have one output neuron with linear activation
                    # for a regression network
                    output_neuron = self._output_layer.get_neurons()[0]
                    z = self._calculate_pre_activated_value(left_layer,
                                                            output_neuron)
                    output_neuron.set_value(z)
                    # Here we exit the method, returning the value of the
                    # output neuron
                    return [z]
                else:
                    for right_neuron in right_layer.get_neurons():
                        # Calculate the desired values for each neuron
                        z = self._calculate_pre_activated_value(left_layer,
                                                                right_neuron)
                        z_output_layer.append(z)

        # Output layer
        # Activates the output layer using Softmax activation
        softmax_vector = self._activate_output_layer(z_output_layer)
        return softmax_vector

    def _calculate_pre_activated_value(self, left_layer: Layer,
                                       right_neuron: Neuron) -> float:
        """Given a `left_layer` and a `right_neuron`, this calculates the
        activation function and value from the `left_layer` and propagates
        this value to the `right_neuron`

        Parameters
        ----------
        left_layer : Layer
            The current `left_layer` in forward propagation
        right_neuron : Neuron
            The current `right_neuron` in forward propagation
        """
        left_neurons = left_layer.get_neurons()
        i, j = right_neuron.get_id()

        # All edges connecting the left_layer to the right_neuron
        edges = self._edges[i - 1][j]

        # Lists of values and weights, with the bias
        o_list = [neuron.get_value() for neuron in left_neurons]
        w_list = [edge.get_weight() for edge in edges]
        bias = right_neuron.get_bias()

        # Return the pre-activated value for this neuron
        return self._transfer(o_list, w_list + [bias])

    def _activate_output_layer(self, z_list: List[float]) -> List[float]:
        """Activates the values from the `output_layer` using the Softmax
        activation function.

        Returns
        -------
        List[float]
            The `List` of softmax probabilities
        """
        self._softmax.normalise(z_list)
        softmax_neurons = self._output_layer.get_neurons()
        softmax_vector = []
        for j, value in enumerate(z_list):
            softmax = self._softmax(value)
            softmax_vector.append(softmax)
            softmax_neurons[j].set_value(softmax)

        return softmax_vector

    def store_gradient_of_loss(self, edge: Edge, target: float, first: bool):
        """Calculates the gradient of the loss function with respect to one
        weight (assigned to the edge) based on the values at edges of future
        layers. One part of the back propagation process.

        edge : Edge
            The `Edge` containing the weight we are interested in
        target : int
            The target value for the final output node for this specific
            datapoint
        first : bool
            Determines whether we find the bias gradient or not
        """
        left_layer_index = edge.get_id()[0]
        right_neuron = edge.get_right_neuron()

        # Value of left neuron and right neuron
        o_left = edge.get_left_neuron().get_value()
        o_right = right_neuron.get_value()

        right_index, row = right_neuron.get_id()

        # Output layer
        if left_layer_index == self._num_hidden_layers:
            if self._regression:
                delta = self._mse_loss.gradient(o_right, target)
            else:
                delta = o_right - int(row == int(target))
            edge.set_delta(delta)
            edge.loss_gradients.append(round(o_left * delta, 8))
            if first:
                right_neuron.bias_gradients.append(delta)

        # Hidden layers
        else:
            next_layer = self._layers[right_index + 1]
            next_edges = [self._edges[right_index][j][row]
                          for j in range(len(next_layer))]
            factor = sum([new_edge.get_weight() * new_edge.get_delta()
                          for new_edge in next_edges])

            # Constant (either +1 or self._leak)
            relu_grad = self._relu.gradient(o_right)
            delta = factor * relu_grad
            edge.set_delta(delta)
            edge.loss_gradients.append(o_left * delta)
            if first:
                right_neuron.bias_gradients.append(delta)

    def back_propagate_weight(self, edge: Edge):
        """Uses the loss gradients of all datapoints (for this specific edge)
        to perform gradient descent and calculate a new weight for this edge.

        edge : Edge
            The `Edge` whose weight we are interested in updating
        """
        current_weight = edge.get_weight()
        batch_size = len(edge.loss_gradients)
        avg_loss_gradient = sum(edge.loss_gradients) / batch_size
        if self._adaptive:
            velocity = (self._gamma * edge.get_velocity()
                        + self._learning_rate * avg_loss_gradient)
            edge.set_weight(current_weight - velocity)
            edge.set_velocity(velocity)
        else:
            edge.set_weight(current_weight - self._learning_rate
                            * avg_loss_gradient)
        edge.loss_gradients = []

    def back_propagate_bias(self, neuron: Neuron):
        """Uses the bias gradients of all datapoints (for this specific neuron)
        to perform gradient descent and calculate a new bias for this neuron.

        neuron : Neuron
            The `Neuron` whose bias we are interested in updating
        """
        current_bias = neuron.get_bias()
        batch_size = len(neuron.bias_gradients)
        avg_bias_gradient = sum(neuron.bias_gradients) / batch_size
        neuron.set_bias(current_bias - self._learning_rate
                        * avg_bias_gradient)
        neuron.bias_gradients = []

    def visualise_network(self, title: str = ''):
        """Uses the networkx package to visualise the network.

        Parameters
        ----------
        title : str [Optional, Default='']
            An optional title for the network.
        """
        graph = nx.Graph()
        for edge_layer in self._edges:
            for right_neuron in edge_layer:
                for edge in right_neuron:
                    left_id = edge.get_left_neuron().get_id()
                    right_id = edge.get_right_neuron().get_id()
                    left_id_str = f"{left_id[0]},{left_id[1]}"
                    right_id_str = f"{right_id[0]},{right_id[1]}"
                    graph.add_node(left_id_str, layer=left_id[0])
                    graph.add_node(right_id_str, layer=right_id[0])
                    graph.add_edge(left_id_str, right_id_str)
        pos = nx.multipartite_layout(graph, subset_key="layer")
        nx.draw_networkx(graph, pos=pos,
                         node_size=480, node_color='green')
        joining_str = f"_{title}" if title else ""
        title_str = f": {title}" if title else ""
        plt.title(f"Network{title_str}")
        plt.savefig(f"network{joining_str}.png")
        plt.clf()

    def get_edges(self) -> List[List[List[Edge]]]:
        """Getter method for edges

        Returns
        -------
        List[List[List[Edge]]]
            A list of `Edges`
        """
        return self._edges

    def get_layers(self) -> List[Layer]:
        """Getter method for layers

        Returns
        -------
        List[Layer]
            A list of `Layers`
        """
        return self._layers

    def get_neuron_counts(self) -> List[int]:
        """Getter method for neuron_counts.

        Returns
        -------
        List[int]
            A list of numbers of neurons per layer
        """
        return ([self._num_features] + self._neuron_counts
                + [len(self._output_layer)])

    def is_regressor(self) -> bool:
        """Getter method for regression.

        Returns
        -------
        bool
            Whether we do regression or classification
        """
        return self._regression

File: javaNeuralNetwork/src/main/java/neural_network/components/Network.java
package neural_network.components;

import neural_network.functions.LeakyReLU;
import neural_network.functions.MSELoss;
import neural_network.functions.Softmax;
import neural_network.functions.TransferFunction;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.stream.Stream;

import static java.lang.Math.sqrt;

/** Class to represent the whole network.
 *
 */
public class Network {

    // Numbers of neurons
    private final int numFeatures;
    private final int numHiddenLayers;
    private final List<Integer> neuronCounts = new ArrayList<>();

    // Regression boolean
    private final boolean regression;

    // Layer info
    private final Layer inputLayer;
    private final List<Layer> hiddenLayers = new ArrayList<>();
    private final Layer outputLayer;
    private final List<Layer> layers = new ArrayList<>();

    // Edge info
    private final List<List<List<Edge>>> edges = new ArrayList<>();

    // Functions
    private final TransferFunction transfer;
    private final LeakyReLU relu;
    private MSELoss mseLoss = null;
    private Softmax softmax = null;

    // Hyperparameters
    private final double learningRate;
    private final boolean adaptive;
    private final double gamma;
    private static Random random = new Random();

    /** Minimal constructor - need to specify number of inputs, outputs and
     * hidden layers.
     *
     * @param numFeatures The number of coordinates per datapoint.
     * @param numHiddenLayers The number of hidden {@code Layers} for the {@code Network}.
     * @param neuronCounts The number of {@code Neurons} in each hidden {@code Layer}.
     */
    public Network(int numFeatures, int numHiddenLayers, List<Integer> neuronCounts) {
        this(numFeatures, numHiddenLayers, neuronCounts, 2,0.01,
                0.01, false, false, Double.NaN,
                false);
    }

    /** Medium level constructor - for classification.
     *
     * @param numFeatures The number of coordinates per datapoint.
     * @param numHiddenLayers The number of hidden {@code Layers} for the {@code Network}.
     * @param neuronCounts The number of {@code Neurons} in each hidden {@code Layer}.
     * @param numClasses The number of classes for a classification task.
     * @param leak The leak for the {@code LeakyReLU}.
     * @param learningRate The learning rate of the {@code Network}.
     */
    public Network(int numFeatures, int numHiddenLayers, List<Integer> neuronCounts,
                   int numClasses, double leak, double learningRate) {
        this(numFeatures, numHiddenLayers, neuronCounts, numClasses, leak,
                learningRate, false, false, Double.NaN, false);
    }

    /** Most general constructor.
     *
     * @param numFeatures The number of coordinates per datapoint.
     * @param numHiddenLayers The number of hidden {@code Layers} for the {@code Network}.
     * @param neuronCounts The number of {@code Neurons} in each hidden {@code Layer}.
     * @param numClasses The number of classes for a classification task.
     * @param leak The leak for the {@code LeakyReLU}.
     * @param learningRate The learning rate of the {@code Network}.
     * @param regression Whether we are performing regression or not ({@code false} for
     *                   classification).
     * @param adaptive Whether we wish to have an adaptive {@code learningRate} or not.
     * @param gamma The adaptive learning rate parameter.
     * @param heWeights Whether we wish to use He initialisation for the weights or not.
     */
    public Network(int numFeatures, int numHiddenLayers, List<Integer> neuronCounts,
                   int numClasses, double leak, double learningRate, boolean regression,
                   boolean adaptive, double gamma, boolean heWeights) {
        if (numHiddenLayers != neuronCounts.size()) {
            throw new IllegalArgumentException(
                    "neuronCounts (%d) must have a length equal to numHiddenLayers (%d)"
                            .formatted(neuronCounts.size(), numHiddenLayers));
        }

        // Initial set up
        this.numFeatures = numFeatures;
        this.numHiddenLayers = numHiddenLayers;
        this.neuronCounts.addAll(neuronCounts);
        this.regression = regression;

        // If we are doing regression, set numClasses to 1 no matter the input value
        if (regression) {
            numClasses = 1;
        }

        // Layers
        this.inputLayer = new Layer(0, numFeatures);
        this.hiddenLayers.addAll(Stream
                .iterate(1, j -> j <= numHiddenLayers, j -> j + 1)
                .map(j -> new Layer(j, neuronCounts.get(j - 1)))
                .toList());
        this.outputLayer = new Layer(numHiddenLayers + 1, numClasses);
        this.layers.add(inputLayer);
        this.layers.addAll(hiddenLayers);
        this.layers.add(outputLayer);

        // Edges
        // Order in edges is left layer then right neuron then left neuron
        for (int i = 0; i <= numHiddenLayers; i ++) {
            Layer leftLayer = layers.get(i);
            Layer rightLayer = layers.get(i + 1);
            List<List<Edge>> layerList = new ArrayList<>();
            for (Neuron rightNeuron : rightLayer.getNeurons()) {
                List<Edge> edgeList = leftLayer.getNeurons().stream()
                        .map(leftNeuron -> new Edge(leftNeuron, rightNeuron))
                        .toList();
                layerList.add(edgeList);

                // If we are using He initialisation, we will set the weights here
                // according to that (mean = 0, std dev = 2 / leftLayer.size())
                if (heWeights) {
                    int n = leftLayer.size();
                    for (Edge edge : edgeList) {
                        edge.setWeight(random.nextGaussian(0.0,
                                sqrt(2.0 / n)));
                    }
                }
            }
            edges.add(layerList);
        }

        // Functions
        this.transfer = new TransferFunction();
        this.relu = new LeakyReLU(leak);
        if (regression) {
            this.mseLoss = new MSELoss();
        } else {
            this.softmax = new Softmax();
        }

        // Hyper-parameters
        this.learningRate = learningRate;
        this.adaptive = adaptive;
        this.gamma = gamma;
    }

    /** Performs a forward pass for one datapoint, excluding the ground
     * truth value. This method returns the predicted value in the final
     * {@code Neuron} in the {@code outputLayer}.
     *
     * @param x The datapoint, with all features.
     * @return The softmax probabilities of each class (for classification) or
     *             the predicted regression value (for regression).
     */
    public List<Double> forwardPassOneDatapoint(List<Double> x) {
        if (x.size() != inputLayer.size()) {
            throw new IllegalArgumentException(
                    "Number of features must match the number of neurons in the input layer " +
                            "(%d != %d)".formatted(x.size(), inputLayer.size()));
        }

        // Input layer
        List<Neuron> inputNeurons = inputLayer.getNeurons();
        for (int j = 0; j < inputLayer.size(); j ++) {
            inputNeurons.get(j).setValue(x.get(j));
        }

        // This is setting up for the output layer
        List<Double> zOutputLayer = new ArrayList<>();
        // Hidden layers and output layer
        for (Layer leftLayer : layers.subList(0, layers.size() - 1)) {
            Layer rightLayer = layers.get(leftLayer.getId() + 1);

            if (rightLayer.getId() != outputLayer.getId()) {
                for (Neuron rightNeuron : rightLayer.getNeurons()) {
                    // Calculates the desired value for each neuron
                    double z = calculatePreActivatedValue(leftLayer, rightNeuron);
                    // Uses ReLU activation for the neuron
                    rightNeuron.setValue(relu.call(z));
                }
            } else {
                if (regression) {
                    // We only have one output neuron with linear activation
                    // for a regression network
                    Neuron outputNeuron = outputLayer.getNeurons().get(0);
                    double z = calculatePreActivatedValue(leftLayer, outputNeuron);
                    outputNeuron.setValue(z);
                    // Here, we exit the method, returning the value of the
                    // output neuron
                    return new ArrayList<>(List.of(z));
                } else {
                    for (Neuron rightNeuron : rightLayer.getNeurons()) {
                        double z = calculatePreActivatedValue(leftLayer, rightNeuron);
                        zOutputLayer.add(z);
                    }
                }
            }
        }

        // Output layer
        // Activates the output layer using softmax activation
        return activateOutputLayer(zOutputLayer);
    }

    /** Given a {@code leftLayer} and a {@code rightNeuron}, this calculates the
     * activation function and value from the {@code leftLayer} and propagates
     * this value to the {@code rightNeuron}.
     *
     * @param leftLayer The current left {@code Layer} in forward propagation.
     * @param rightNeuron The current right {@code Neuron} in forward propagation.
     * @return The value returned by the {@code transferFunction} before activation.
     */
    double calculatePreActivatedValue(Layer leftLayer, Neuron rightNeuron) {
        List<Neuron> leftNeurons = leftLayer.getNeurons();
        int i = rightNeuron.getId().get(0); int j = rightNeuron.getId().get(1);

        // All edges connecting the leftLayer to the rightNeuron
        List<Edge> innerEdges = edges.get(i - 1).get(j);

        // Lists of values and weights, with the bias
        List<Double> oList = leftNeurons.stream()
                .map(Neuron::getValue)
                .toList();
        List<Double> weights = innerEdges.stream()
                .map(Edge::getWeight)
                .toList();
        double bias = rightNeuron.getBias();
        transfer.bindWeights(weights);
        transfer.bindBias(bias);

        // Return the pre-activated value for this neuron
        return transfer.call(oList);
    }

    /** Activates the values from the {@code outputLayer} using the {@code softmax}
     * activation function.
     *
     * @param zOutputLayer The pre-activated values from the output layer in a
     *                     classification problem.
     * @return The list of softmax probabilities.
     */
    List<Double> activateOutputLayer(List<Double> zOutputLayer) {
        softmax.normalise(zOutputLayer);
        List<Neuron> softmaxNeurons = outputLayer.getNeurons();
        List<Double> softmaxVector = new ArrayList<>();
        for (int j = 0; j < zOutputLayer.size(); j ++) {
            double softmaxValue = softmax.call(zOutputLayer.get(j));
            softmaxVector.add(softmaxValue);
            softmaxNeurons.get(j).setValue(softmaxValue);
        }
        return softmaxVector;
    }

    /** Calculates the gradient of the loss function with respect to one
     * weight (assigned to the {@code edge}) based on the values at edges of future
     * layers. One part of the back propagation process.
     *
     * @param edge The {@code Edge} containing the weight we are interested in.
     * @param target The target value for the final output neuron for this specific
     *               datapoint.
     * @param first Determines whether we find the bias gradient or not.
     */
    public void storeGradientOfLoss(Edge edge, double target, boolean first) {
        int leftLayerIndex = edge.getId().get(0);
        Neuron rightNeuron = edge.getRightNeuron();

        // Value of left neuron and right neuron
        double oLeft = edge.getLeftNeuron().getValue();
        double oRight = rightNeuron.getValue();

        int rightIndex = rightNeuron.getId().get(0); int row = rightNeuron.getId().get(1);

        // Output layer
        if (leftLayerIndex == numHiddenLayers) {
            double delta = (regression) ? mseLoss.gradient(oRight, target)
                    : oRight - Boolean.compare(row == (int) target, false);
            edge.setDelta(delta);
            edge.addLossGradient(oLeft * delta);

            // If this is the first time we call this function for the neuron,
            // we need to also store the gradient of the loss for the bias
            if (first) {
                rightNeuron.addBiasGradient(delta);
            }
        } else {
            // Hidden layers
            Layer nextLayer = layers.get(rightIndex + 1);

            // Edges connected to the current rightNeuron in the layer to
            // the right of the right layer
            List<Edge> nextEdges = Stream
                    .iterate(0, j -> j < nextLayer.size(), j -> j + 1)
                    .map(j -> edges.get(rightIndex).get(j).get(row))
                    .toList();

            double factor = nextEdges.stream()
                    .mapToDouble(newEdge -> newEdge.getWeight() * newEdge.getDelta())
                    .sum();

            // Constant (either +1 or self._leak)
            double reluGrad = relu.gradient(oRight);
            double delta = factor * reluGrad;
            edge.setDelta(delta);
            edge.addLossGradient(oLeft * delta);

            if (first) {
                rightNeuron.addBiasGradient(delta);
            }
        }
    }

    /** Uses the loss gradients of all datapoints (for this specific {@code edge})
     * to perform gradient descent and calculate a new weight for this {@code edge}.
     *
     * @param edge The {@code Edge} whose weight we are interested in updating.
     */
    public void backPropagateWeight(Edge edge) {
        double currentWeight = edge.getWeight();

        // The number of datapoints which we have passed through the network
        // in a batch
        int batchSize = edge.getLossGradients().size();
        double avgLossGradient = edge.getLossGradients().stream()
                .mapToDouble(a -> a)
                .sum() / batchSize;

        // If the network is adaptive we need a different gradient descent algorithm
        if (adaptive) {
            double velocity = gamma * edge.getVelocity() +
                    learningRate * avgLossGradient;
            edge.setWeight(currentWeight - velocity);
            edge.setVelocity(velocity);
        } else {
            edge.setWeight(currentWeight - learningRate * avgLossGradient);
        }

        // Reset for the next batch
        edge.clearLossGradients();
    }

    /** Uses the bias gradients of all datapoints (for this specific {@code Neuron})
     * to perform gradient descent and calculate a new bias for this {@code Neuron}.
     *
     * @param neuron The {@code Neuron} whose bias we are interested in updating.
     */
    public void backPropagateBias(Neuron neuron) {
        double currentBias = neuron.getBias();
        double batchSize = neuron.getBiasGradients().size();
        double avgBiasGradient = neuron.getBiasGradients().stream()
                .mapToDouble(a -> a)
                .sum() / batchSize;
        neuron.setBias(currentBias - learningRate * avgBiasGradient);
        neuron.clearBiasGradients();
    }

    /** Getter method for all the neuron counts, for the input,
     * hidden and output {@code Layers}.
     *
     * @return All neuron counts.
     */
    public List<Integer> getNeuronCounts() {
        List<Integer> allNeuronCounts = new ArrayList<>();
        allNeuronCounts.add(numFeatures);
        allNeuronCounts.addAll(neuronCounts);
        allNeuronCounts.add(outputLayer.size());
        return allNeuronCounts;
    }

    /** Getter method for the {@code regression} parameter.
     *
     * @return {@code true} if we are doing regression, {@code false} otherwise.
     */
    public boolean isRegressor() {
        return regression;
    }

    /** Getter method for the {@code layers}.
     *
     * @return A list of all the layers of the {@code Network}.
     */
    public List<Layer> getLayers() {
        return List.copyOf(layers);
    }

    /** Getter method for the {@code edges}.
     *
     * @return A list of all the edges of the {@code Network}.
     */
    public List<List<List<Edge>>> getEdges() {
        List<List<List<Edge>>> copyEdges = new ArrayList<>();
        for (List<List<Edge>> list1 : edges) {
            List<List<Edge>> copyEdges1 = new ArrayList<>();
            for (List<Edge> list2 : list1) {
                copyEdges1.add(List.copyOf(list2));
            }
            copyEdges.add(copyEdges1);
        }
        return copyEdges;
    }

    /** Setter method for {@code random}.
     *
     * @param random The new {@code random}. This can be used for mocking
     *               purposes.
     */
    static void setRandom(Random random) {
        Network.random = random;
    }
}
